mm2src/lp_ordermatch.rs

```diff
@@ -580,55 +518,65 @@ impl From<Box<trie_db::TrieError<H64, sp_trie::Error>>> for TrieDiffHistoryError
     }
 }
 
+fn get_full_trie<Key, Value>(
+    trie_root: &H64,
+    db: &MemoryDB<Blake2Hasher64>,
+) -> Result<Vec<(Key, Value)>, TrieDiffHistoryError>
+where
+    Key: Clone + Eq + std::hash::Hash + TryFromBytes,
+    Value: Clone + TryFromBytes,
+{
+    let trie = TrieDB::<Layout>::new(db, trie_root)?;
+    let trie: Result<Vec<_>, TrieDiffHistoryError> = trie
+        .iter()?
+        .map(|key_value| {
+            let (key, value) = key_value?;
+            Ok((TryFromBytes::try_from_bytes(key)?, TryFromBytes::try_from_bytes(value)?))
+        })
+        .collect();
+    trie
+}
+
 impl<Key: Clone + Eq + std::hash::Hash + TryFromBytes, Value: Clone + TryFromBytes> DeltaOrFullTrie<Key, Value> {
     fn from_history(
         history: &TrieDiffHistory<Key, Value>,
         from_hash: H64,
-        trie_from_hash: H64,
+        actual_trie_root: H64,
         db: &MemoryDB<Blake2Hasher64>,
     ) -> Result<DeltaOrFullTrie<Key, Value>, TrieDiffHistoryError> {
-        match history.get(&from_hash) {
-            Some(delta) => {
-                let mut current_delta = delta;
-                let mut total_delta = HashMap::new();
-                for (key, new_value) in &delta.delta {
+        if let Some(delta) = history.get(&from_hash) {
+            let mut current_delta = delta;
+            let mut total_delta = HashMap::new();
+            for (key, new_value) in &delta.delta {
+                total_delta.insert(key.clone(), new_value.clone());
+            }
+            while let Some(cur) = history.get(&current_delta.next_root) {
+                current_delta = cur;
+                for (key, new_value) in &current_delta.delta {
                     total_delta.insert(key.clone(), new_value.clone());
                 }
-                while let Some(cur) = history.get(&current_delta.next_root) {
-                    current_delta = cur;
-                    for (key, new_value) in &current_delta.delta {
-                        total_delta.insert(key.clone(), new_value.clone());
-                    }
-                }
-                Ok(DeltaOrFullTrie::Delta(total_delta))
-            },
-            None => {
-                let trie = TrieDB::<Layout>::new(db, &trie_from_hash)?;
-                let trie: Result<Vec<_>, TrieDiffHistoryError> = trie
-                    .iter()?
-                    .map(|key_value| {
-                        let (key, value) = key_value?;
-                        Ok((TryFromBytes::try_from_bytes(key)?, TryFromBytes::try_from_bytes(value)?))
-                    })
-                    .collect();
-                Ok(DeltaOrFullTrie::FullTrie(trie?))
-            },
+            }
+            if current_delta.next_root == actual_trie_root {
```

Comment:
Also I added this checking if the last history changes doesn't bring us to the actual trie_root.

Reply:
If we hit some state that comes after `actual_trie_root` we can add last signed message that has the new root and provide it to requesting node as proof of state change.
We can postpone it, but last signed message is essential for this and it will hard to add later. Could you please add `last_signed_message` to the `SyncPubkeyOrderbookStateRes`?